identity_core/src/crypto/merkle_tree/proof.rs

```diff
@@ -0,0 +1,47 @@
+use core::fmt::Debug;
+use core::fmt::Formatter;
+use core::fmt::Result;
+use digest::Digest;
+use subtle::ConstantTimeEq;
+
+use crate::crypto::merkle_tree::Hash;
+use crate::crypto::merkle_tree::Node;
+
+/// An inclusion proof that allows proving the existence of data in a
+/// [`Merkle tree`](`struct@super::MTree`).
+pub struct Proof<D: Digest> {
+    nodes: Box<[Node<D>]>,
+}
+
+impl<D: Digest> Proof<D> {
```

Comment:
Just to make sure, would you be able to verify a proof with only the hashes that you need to recreate the top hash? Without having to know ALL leave hashes, right? ;D

Reply:
Yes, the type `Node<D>` here is just a hash with a prepended tag byte and the `Proof<D>` is simply the list of sibling nodes from the target leaf to the root - computing and verifying the root hash from this list is simple:

```rust
let computed_root = proof_nodes
  .iter()
  .fold(target_hash, |acc, item| match item {
    Node::L(hash) => hash_branch(hash, acc),
    Node::R(hash) => hash_branch(acc, hash),
  });

computed_root.ct_eq(target_root);
```